neuro_evolution_ctrnn/optimizer/optimizer_mu_lambda.py

```diff
@@ -65,35 +68,82 @@ def fct_mutation_learned(ind1):
-            toolbox.population = self.toolbox.population(n=self.conf.mu)
-            toolbox.logbook = self.create_logbook(conf)
-            toolbox.recorded_individuals = []
+            # TODO get_checkpoint(from_checkpoint) is already called in super()
+            self.population = get_checkpoint(from_checkpoint)["population"]
+        self.clone = deepcopy
 
     def train(self, number_generations) -> tools.Logbook:
-        return algorithms.eaMuPlusLambda(toolbox=self.toolbox, ngen=number_generations)
+        verbose = __debug__
+        gen = None
+        current_seed = None
+
+        for gen in range(self.initial_generation, number_generations):
+            self.record_individuals(self.population)
+            extra = []
+            if self.hof.items:
+                extra = list(map(self.clone, random.sample(self.population, self.conf.extra_from_hof)))
+            offspring = varOr(self.population + extra, self, self.conf.lambda_, 1 - self.conf.mutpb,
+                              self.conf.mutpb)
+
+            for ind in offspring:
+                # just a little extra for when we later want to analyze the hof manually
+                ind.generation = gen
+
+            for ind in self.population:
+                del ind.fitness.values
+            candidates = self.population + offspring
+
+            nevals, total_steps, current_seed, candidates = self.evaluate_candidates(candidates)
+            if self.hof is not None:
+                self.hof.update(offspring)
+
+            record = self.stats.compile(candidates) if self.stats is not None else {}
+            self.population[:] = self.select(candidates, self.conf.mu)
+
+            validation_fitness, validation_steps = self.validation_runs(self.population)
+            self.logbook.record(gen=gen, nevals=nevals, steps=total_steps + validation_steps,
+                                val_fit=validation_fitness, **record)
+            if verbose:
+                print(self.logbook.stream)
+            self.write_checkpoint(data=dict(generation=gen, halloffame=self.hof, population=self.population,
+                                            logbook=self.logbook, last_seed=current_seed, strategy=None,
+                                            recorded_individuals=self.recorded_individuals))
+        self.final_checkpoint_data = dict(generation=gen, halloffame=self.hof, population=self.population,
+                                          logbook=self.logbook, last_seed=current_seed, strategy=None,
+                                          recorded_individuals=self.recorded_individuals)
+        return self.logbook
+
+    @classmethod
+    def strip_strategy_from_population(cls, population, strategy_parameter_per_gene=False):
+        if len(population) == 0:
+            return population
+        return list(np.array(population)[:,
+                    :-(cls.get_number_genes_for_strategy(len(population[0]), strategy_parameter_per_gene))])
```

Comment:
Die Methode ist ziemlich unübersichtlich, was u.A. daran liegen dürfte, dass meine ursprüngliche Implementierung davon eher mangelhaft war

Der Wert für `strategy_parameter_per_gene` ist für den optimizer immer gleich, oder? Man könnte den zu einer member variable machen. Das gleiche für `len(population[0]`. Und dann liefert `get_number_genes_for_strategy` auch immer den gleichen wert zurück, oder? Dann kann man das anstatt zu einer methode auch zu eine Variable machen.

Der Fall mit  `len(population) == 0` trifft nur im Fehlerfall ein, oder? Wenn ja, dann  entweder eine Exception werfen, oder das `if` ganz  weglassen (und darauf vertrauen, dass der fehler dann irgendwo anders geworfen wird). Dieses stillschweigende Fehler-Beheben ist auf jeden Fall immer schlecht. Weil wenn der Fehler dann auftritt, dann wird das experiment sowieso ganz gegen die wand laufen. Und wenn dann viele solche stellen drin sind, wo stillschweigende Fehler "gefixt" werden, dann wird das troubleshooting viel viel schwieriger. 



```suggestion
    def strip_strategy_from_population(self, population):
        return list(np.array(population)[:, :-self.number_strategy_params_in_genome])
```